src/connector/src/sink/big_query.rs

```diff
@@ -82,23 +87,116 @@ pub struct BigQueryCommon {
     pub dataset: String,
     #[serde(rename = "bigquery.table")]
     pub table: String,
-    #[serde(rename = "bigquery.max_batch_rows", default = "default_max_batch_rows")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub max_batch_rows: usize,
-    #[serde(rename = "bigquery.retry_times", default = "default_retry_times")]
-    #[serde_as(as = "DisplayFromStr")]
-    pub retry_times: usize,
     #[serde(default)] // default false
     #[serde_as(as = "DisplayFromStr")]
     pub auto_create: bool,
 }
 
-fn default_max_batch_rows() -> usize {
-    1024
+struct BigQueryFutureManager {
+    offset_queue: VecDeque<TruncateOffset>,
+    resp_stream: Pin<Box<dyn Stream<Item = Result<()>> + Send>>,
+    max_future_num: usize,
+    tx: mpsc::Sender<()>,
+    rx: mpsc::Receiver<()>,
+}
+impl BigQueryFutureManager {
+    pub fn new(
+        max_future_num: usize,
+        resp_stream: impl Stream<Item = Result<()>> + Send + 'static,
+    ) -> Self {
+        let offset_queue = VecDeque::with_capacity(max_future_num);
+        let (tx, rx) = mpsc::channel(1);
+        Self {
+            offset_queue,
+            resp_stream: Box::pin(resp_stream),
+            max_future_num,
+            tx,
+            rx,
+        }
+    }
+
+    pub async fn add_offset(&mut self, offset: TruncateOffset) -> Result<()> {
+        self.offset_queue.push_back(offset);
+        let _ = self.tx.try_send(());
+        while self.offset_queue.len() >= self.max_future_num {
+            self.wait_one_offset().await?;
+        }
+        Ok(())
+    }
+
+    pub async fn wait_next_offset(&mut self) -> Result<TruncateOffset> {
+        loop {
+            if let Some(offset) = self.wait_one_offset().await? {
+                return Ok(offset);
+            } else {
+                self.rx.recv().await.unwrap();
```

Comment:
The `tx` and `rx` is not needed. When we await on `self.rx.recv()`, we have the ownership of `self`, and no one will be able to access `self.tx` to notify anything.

The future manager can be very simple. It only needs two methods
```
fn add_offset(&mut self, offset: TruncateOffset) {
    self.offset_queue.push_back(offset);
}

fn wait_next_offset(&mut self) -> Result<TruncateOffset> {
    if let Some(TruncateOffset::Barrier(_)) = self.offset_queue.front() {
        return Ok(self.offset_queue.pop_front().unwrap());
    }
    let resp = self.resp_stream.try_next().await?.ok_or_else("end of stream")?;
    self.offset_queue.pop_front().ok_or_else("should have pending chunk offset when we receive new response")
}
```

We can limit the number of pending request by not polling the log reader in the `select!` by
```
select!(
    self.future_manager.wait_next_offset() => { ... }
    self.log_reader.next_item() if self.future_manager.offset_queue.len() <= MAX_XXX => { ... }
)
```